### PR TITLE
Track additional characteristics in Scenario

### PR DESCRIPTION
**Pull request by AaronGullickson**
This PR refactors some features out of AtBScenario into Scenario and adds some additional features to Scenario. This PR is tied to the Story Arc PR #2997, and should allow creators to create pre-made scenarios for story arcs. However, it is generalizable enough that it could have other potential uses as well. 

## Current State

I believe this PR is ready for review. I am going to look and see what I can clean up based on the automated reports, but feel free to give feedback at this point.

## Refactoring

The following variables and related methods were re-factored out of AtBScenario to Scenario:

1. `botForces` and `botForcesStubs`. This allows basic scenarios to include pre-loaded OpFor and Allied units. I also borrowed code from AtBGameThread and inserted it into GameThread so that these botForces are actually loaded. At a basic level, this feature can include a fixed set of entities listed in the BotForce object, but I also added an additional class described below that can do randomization of botForces. There is another variable in AtBScenario for allies (alliesPlayer) but I did not refactor this variable because allies can be created in BotForce simply enough by assigning them to team 1 (always the player's team). 
2. Map generation variables, specifically `terrainType`, `mapSizeX`, `mapSizeY`, `map`, and `usingFixedMap`. These function identically as they did in AtBScenario. 
3. PlanetaryConditions variables. I also added additional variables for temperature, electro-magnetic variables, and blowing sand. These variables are not currently used by AtBScenario but could be added.
4. The `start` variable that gives the starting deployment position of the player's units.

I don't *believe* any of these changes will affect AtB functioning at all because AtBScenario inherits from Scenario and I made sure to keep methods the same.  I did have to give some of the variables protected rather than private access because the initialize method in AtBScenario assigns some of these variables directly.

I also did not need to do anything to ScenarioObjectives. These were already a part of the basic Scenario object and they can be used more or less directly, although I did need to make a few changes to make sure they were checked for in basic scenarios and not just AtB games. 

## GUI changes

I made a variety of changes to `ScenarioViewPanel` so that all of these new features in scenario will actually show up to the player. I borrowed quite a bit from `AtBScenarioViewPanel` but `ScenarioViewPanel` involves no dynamic buttons just information.  I also did a *lot* of cleanup on the code there to make it more compact and produce a more pleasing display. Here is an example of the new look: 

Screen Shot 2021-12-27 at 4 11 29 PM

### BotForceRandomizer

I wanted people to be able to create randomized BotForces as well as fixed ones. I also wanted these to be able to scaled to variable player deployments. I originally thought I could just use or rework the code from `ScenarioForceTemplate` and `AtBDynamicScenarioFactory`. However the logic of these objects is quite complex and difficult to separate from the expectations of AtB game play. 

So, instead, I created a new class called `BotForceRandomizer` that can be added to a `BotForce` as a variable. It is a lightweight object that just tracks some parameters for random generation such as faction, quality, unit type, and skill level. The code for generating units borrows substantially from code in `AtBDynamicScenarioFactory` but is much, much simpler. Part of the reason for this is that, players/creators can mix fixed unit lists and random forces in BotForces. So rather than try to consider all the very specific cases that AtB needs to consider like artillery, aeros with bombs, dropship support, etc., this class is set up so that creators can specify special stuff like that in fixed units and then just roll for the remaining units. 

I currently have two scaling techniques which are specified in a private enum. The first is just BV. The second is an "adjusted weight" scaling. This is basically just the total weight of units, but those weights are "weighted" by unit type. So tank weight only counts for 60% of mech weight and so forth. This is loosely based on some old point based systems used in Battletech products. The idea is that this balances forces based on rough unit type and weight parity but does not adjust for skill level, tech level, etc. Right now the unit type weights used are hard coded but this could be made user editable in the future. It should also be easy based on the way I have this set up to add other scaling methods in the future. 

Right now the generator works pretty decently, but there is definitely room for improvement. My initial goal here is just to get the architecture set up and then tinker with the details later based on how it plays.

### ScenarioDeploymentLimit

This feature enables limits on what units the player can deploy to a scenario. All of the parameters for this leave in the ScenarioDeploymentLimit class and an instance of this class can optionally be added to Scenario. If it is not present, then limits are not enforced. Limits are enforced in the following way.

* allowedUnitTypes - This List defines what unit types can be deployed to the scenario. If it is empty, then all unit types are accepted.
* quantityLimit. This variable places a maximum value on units that can be deployed. That maximum limit is flexible depending on the value of QuantityType and CountType enums in ScenarioDeploymentLimit. QuantityType can be PERCENT or ABSOLUTE. If PERCENT is used then quantityLimit is treated as a percent of the player's valid forces for the scenario. So you could set the limit to be 50% of the player's forces for example. What exactly quantityLimit measures can be changed by CountType which can be BV or UNIT. If BV, then quantity limits are based on BV. If UNIT then the limit is number of units. So, quantityLimit=50, QuantityType=PERCENT, and CountType=UNIT would allow the player to field half of their eligible units to the scenario.
* requiredPersonnel - this can be used to require certain personnel in the scenario by their UUID
* requiredUnits - just like requiredPersonnel but for units

Most of these limits are enforced by the `canDeployForces` and `canDeployUnits` methods in Scenario, which I refactored from AtBScenario. The methods in AtBScenario are unchanged and just override the super methods. I also added a new Scenario#canStartScenario method that is checked in the GUI to see if all of the scenario buttons should be enabled, which helped to clean up some real messy code. Again AtBScenario has an override method for this that adds one additional condition.

### CODE REVIEW

File: MekHQ/src/mekhq/campaign/mission/ScenarioDeploymentLimit.java [hunk 1]
```diff
@@ -0,0 +1,469 @@
...
+            pw1.println(MekHqXmlUtil.indentStr(indent+1) + "<allowedUnitTypes>");
+            for(int type : allowedUnitTypes) {
+                pw1.println(MekHqXmlUtil.indentStr(indent+2)
+                        +"<allowedUnitType>"
+                        +type
+                        +"</allowedUnitType>");
+            }
+            pw1.println(MekHqXmlUtil.indentStr(indent+1) + "</allowedUnitTypes>");
+        }
+        pw1.println(MekHqXmlUtil.indentStr(indent) + "</scenarioDeploymentLimit>");
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/mission/ScenarioDeploymentLimit.java:428]: This needs to be converted to the latest MekHqXmlUtil methods, plus the spacing needs fixing.

File: MekHQ/src/mekhq/campaign/mission/ScenarioDeploymentLimit.java [hunk 2]
```diff
@@ -0,0 +1,469 @@
...
+                    retVal.countType = CountType.valueOf(wn2.getTextContent().trim());
+                } else if (wn2.getNodeName().equalsIgnoreCase("requiredPersonnel")) {
+                    retVal.requiredPersonnel.add(UUID.fromString(wn2.getTextContent().trim()));
+                }  else if (wn2.getNodeName().equalsIgnoreCase("requiredUnit")) {
+                    retVal.requiredUnits.add(UUID.fromString(wn2.getTextContent().trim()));
+                }
+
+            }
+        }  catch (Exception ex) {
+            LogManager.getLogger().error(ex);
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/mission/ScenarioDeploymentLimit.java:463]:
```suggestion
            LogManager.getLogger().error("", ex);
```

**Review by Windchild292 — CHANGES_REQUESTED** [review]
In general this looks good. However, there's a bunch of comments with changes of varying severity, with the largest and most significant being issues in MekHqXmlUtil usage and File I/O in general. Outside of that the comments are largely related to missing brackets, spacing, other text accessibility, and some light code enhancements.

Also, there's a lot of places in BotForceRandomizer and ScenarioDeploymentLimit that could be converted to use streams to improve their readability, but that's just a personal opinion.

File: MekHQ/src/mekhq/GameThread.java [hunk 3]
```diff
@@ -129,6 +194,31 @@ public void run() {
                 client.sendPlayerInfo();
             }
 
+            /* Add bots */
```
[the remark this answers is not in the document]
Comment (AaronGullickson) [reply MekHQ/src/mekhq/GameThread.java:197]: That was just a copy-paste job from AtBThread and probably goes way back. I know that I have used those in the past that way, but it is probably better to use a single line comment. I am happy to change it.

File: MekHQ/src/mekhq/campaign/mission/AtBScenario.java [hunk 4]
```diff
@@ -1833,24 +1706,8 @@ protected void loadFieldsFromXmlNode(final Node wn, final Version version, final
...
-                    if (bf != null) {
-                        addBotForce(bf);
-                    }
                 } else if (wn2.getNodeName().equalsIgnoreCase("alliesPlayerStub")) {
-                    alliesPlayerStub = getEntityStub(wn2);
-                } else if (wn2.getNodeName().equalsIgnoreCase("botForceStub")) {
-                    String name = MekHqXmlUtil.unEscape(wn2.getAttributes().getNamedItem("name").getTextContent());
-                    List<String> stub = getEntityStub(wn2);
-                    botForceStubs.add(new BotForceStub(name, stub));
+                    alliesPlayerStub = Scenario.getEntityStub(wn2);
```
[the remark this answers is not in the document]
Comment (AaronGullickson) [reply MekHQ/src/mekhq/campaign/mission/AtBScenario.java:1710]: Its a static method so I think it needs to be there right?

[the remark this answers is not in the document]
Comment (AaronGullickson) [reply MekHQ/src/mekhq/campaign/mission/AtBScenario.java:1710]: Ok, never mind - it does not.

File: MekHQ/src/mekhq/campaign/mission/BotForceRandomizer.java [hunk 5]
```diff
@@ -0,0 +1,586 @@
...
+     * This function samples from the given gamma distribution to get a random weight class. Results are trimmed
+     * to reasonable values and rounded to integers.
+     * @param gamma The GammaDistribution from which a random value is drawn
+     * @return and integer giving the sampled weight class
+     */
+    private int sampleWeightClass(GammaDistribution gamma) {
+        double weightClass = gamma.sample();
+        weightClass = (weightClass > EntityWeightClass.WEIGHT_ASSAULT) ? EntityWeightClass.WEIGHT_ASSAULT : weightClass;
+        weightClass = (weightClass < EntityWeightClass.WEIGHT_LIGHT) ? EntityWeightClass.WEIGHT_LIGHT : weightClass;
+        return (int) Math.round(weightClass);
```
[the remark this answers is not in the document]
Comment (AaronGullickson) [reply MekHQ/src/mekhq/campaign/mission/BotForceRandomizer.java:356]: That makes sense, changing it to the following in the forthcoming commit:

```java
private int sampleWeightClass(GammaDistribution gamma) {
        int weightClass = (int) Math.round(gamma.sample());
        // clamp to weight limits
        weightClass = Math.max(EntityWeightClass.WEIGHT_LIGHT, Math.min(EntityWeightClass.WEIGHT_ASSAULT, weightClass));
        return weightClass;
    }
```

File: MekHQ/src/mekhq/campaign/mission/BotForceRandomizer.java [hunk 6]
```diff
@@ -0,0 +1,586 @@
...
+     */
+    private double calculateMeanWeightClass(List<Unit> playerUnits) {
+        int sumWeightClass = 0;
+        int nUnits = 0;
+        for (Unit u : playerUnits) {
+            sumWeightClass += u.getEntity().getWeightClass();
+            nUnits += 1;
+        }
+
+        if (nUnits == 0 | sumWeightClass == 0) {
```
[the remark this answers is not in the document]
Comment (AaronGullickson) [reply MekHQ/src/mekhq/campaign/mission/BotForceRandomizer.java:467]: Should have been double - I code more in R so sometimes forget.

File: MekHQ/src/mekhq/campaign/mission/ScenarioDeploymentLimit.java [hunk 7]
```diff
@@ -0,0 +1,469 @@
...
+    public boolean isAllowedType(int unitType) {
+        if (allowedUnitTypes.isEmpty()) {
+            //allow all units if nothing specified
+            return true;
+        }
+        for (int validType : allowedUnitTypes) {
+            if (validType == unitType) {
+                return true;
+            }
+        }
```
[the remark this answers is not in the document]
Comment (AaronGullickson) [reply MekHQ/src/mekhq/campaign/mission/ScenarioDeploymentLimit.java:151]: Ahh, that would make it simpler.

File: MekHQ/src/mekhq/campaign/mission/ScenarioDeploymentLimit.java (hunk 1 above)
Comment (AaronGullickson) [reply MekHQ/src/mekhq/campaign/mission/ScenarioDeploymentLimit.java:428]: Can you give me an example of a class using the latest methods?

File: MekHQ/src/mekhq/gui/view/ScenarioViewPanel.java [hunk 8]
```diff
@@ -44,152 +55,460 @@
...
+    private JPanel pnlOtherForces;
+    private JPanel pnlReport;
+    private JTextPane txtDesc;
 
     private StubTreeModel forceModel;
 
-    public ScenarioViewPanel(Scenario s, Campaign c) {
+    ResourceBundle resourceMap;
+
+    public ScenarioViewPanel(Scenario s, Campaign c, JFrame f) {
```
[the remark this answers is not in the document]
Comment (AaronGullickson) [reply MekHQ/src/mekhq/gui/view/ScenarioViewPanel.java:79]: Done in pending commit

File: MekHQ/src/mekhq/gui/view/ScenarioViewPanel.java [hunk 9]
```diff
@@ -44,152 +55,460 @@
...
-    }
 
-    private void fillStats() {
+        if (!scenario.getScenarioObjectives().isEmpty()) {
+            fillObjectives();
+            pnlObjectives.setAlignmentX(Component.LEFT_ALIGNMENT);
+            add(pnlObjectives);
+        }
+
+        if (scenario.getLoot().size() > 0) {
```
[the remark this answers is not in the document]
Comment (AaronGullickson) [reply MekHQ/src/mekhq/gui/view/ScenarioViewPanel.java:136]: Done in forthcoming commit

File: MekHQ/src/mekhq/gui/view/ScenarioViewPanel.java [hunk 10]
```diff
@@ -44,152 +55,460 @@
...
+        forceTree.setRootVisible(false);
+        pnlForces.add(forceTree, BorderLayout.CENTER);
+
+        pnlOtherForces = new JPanel();
+        pnlOtherForces.setLayout(new BoxLayout(pnlOtherForces, BoxLayout.PAGE_AXIS));
+        pnlOtherForces.setBorder(BorderFactory.createCompoundBorder(
+                BorderFactory.createEmptyBorder(0, 0, 10, 0),
+                BorderFactory.createTitledBorder(resourceMap.getString("pnlOtherForces.title"))));
+
+        for (int i = 0; i < botStubs.size(); i++) {
```
[the remark this answers is not in the document]
Comment (AaronGullickson) [reply MekHQ/src/mekhq/gui/view/ScenarioViewPanel.java:191]: Changed in forthcoming commit

File: MekHQ/src/mekhq/gui/view/ScenarioViewPanel.java [hunk 11]
```diff
@@ -44,152 +55,460 @@
...
+                BorderFactory.createEmptyBorder(0, 0, 10, 0),
+                BorderFactory.createTitledBorder(resourceMap.getString("pnlDeployment.title"))));
+
+        GridBagConstraints leftGbc = new GridBagConstraints();
+        leftGbc.gridx = 0;
+        leftGbc.gridy = 0;
+        leftGbc.gridwidth = 1;
+        leftGbc.weightx = 0.0;
+        leftGbc.weighty = 0.0;
+        leftGbc.insets = new java.awt.Insets(0, 0, 5, 10);
```
[the remark this answers is not in the document]
Comment (AaronGullickson) [reply MekHQ/src/mekhq/gui/view/ScenarioViewPanel.java:248]: Yeah, thats from some copy-paste of old code. I tried to clean up that stuff but missed a few.

File: MekHQ/src/mekhq/gui/view/ScenarioViewPanel.java [hunk 12]
```diff
@@ -44,152 +55,460 @@
...
+        pnlObjectives.setBorder(BorderFactory.createCompoundBorder(
+                BorderFactory.createEmptyBorder(0, 0, 10, 0),
+                BorderFactory.createTitledBorder(resourceMap.getString("pnlObjectives.title"))));
+
+        StringBuilder objectiveBuilder = new StringBuilder();
+
+        for (ScenarioObjective objective : scenario.getScenarioObjectives()) {
+            objectiveBuilder.append("### ");
+            objectiveBuilder.append(objective.getDescription());
+            objectiveBuilder.append("  \n");
```
[the remark this answers is not in the document]
Comment (AaronGullickson) [reply MekHQ/src/mekhq/gui/view/ScenarioViewPanel.java:314]: Its for the markdown syntax. Two spaces at the end of a line is the signifier to start a new line but not a new paragraph.

**Comment by AaronGullickson** [comment]
Thanks for the very detailed review @Windchild292. I have made all of the requested changes except the two cases of modernizing the XML file I/O. Can you tell me a class that is fully modernized so I can use it as an example?

**Comment by AaronGullickson** [comment]
> Thanks for the very detailed review @Windchild292. I have made all of the requested changes except the two cases of modernizing the XML file I/O. Can you tell me a class that is fully modernized so I can use it as an example?

Never mind I think I fixed this in the latest commit.

**Review by Windchild292 — CHANGES_REQUESTED** [review]
Just a few remaining spacing/brackets issues, then this will be good to merge

File: MekHQ/src/mekhq/campaign/mission/ScenarioDeploymentLimit.java [hunk 13]
```diff
@@ -320,7 +312,7 @@ public String getRequiredPersonnelDesc(Campaign c) {
         ArrayList<String> personNames = new ArrayList<String>();
```
[the remark this answers is not in the document]
Comment (AaronGullickson) [reply MekHQ/src/mekhq/campaign/mission/ScenarioDeploymentLimit.java:312]: Yeah, I noticed some automatic comments on that as well. I think I did it in a few places, but I can try to clear them out.

**Comment by AaronGullickson** [comment]
Ok @Windchild292. Think I got all of them.

**Review by Windchild292 — APPROVED** [review]

**Comment by Windchild292** [comment]
Approved pending CI (just for safety)

**Pull request merged**